src/core/pool.rs

```diff
@@ -146,4 +177,45 @@ mod tests {
             assert_eq!(*pool.acquire(), 11);
         }
     }
+
+    #[test]
+    fn test_queue() {
+        let q = Queue::new();
+        let elem = 5;
+        q.push(elem);
+        let res = q.pop();
+        assert_eq!(res.unwrap(), elem);
+    }
+
+    #[test]
+    fn test_queue_pushes_and_pops_order() {
+        let elements = vec![1, 2, 3, 4];
+        let q = Queue::new();
+        for el in elements.iter() {
+            q.push(el);
+        }
+        for elem in elements.iter() {
+            let popped_el = q.pop().unwrap();
+            assert_eq!(popped_el, elem);
+        }
+    }
+
+    // #[test]
+    //     fn test_pool_dont_panic_on_empty_pop() {
+    //         use std::{thread, time};
+    //         let pool = Pool::new();
+    //         let elements_for_pool = vec![1, 2];
+    //         // Clone vector, so we can calculate its length later
+    //         pool.publish_new_generation(elements_for_pool.clone());
+
+    //         let mut threads = vec![];
+    //         let sleep_dur = time::Duration::from_millis(10);
+    //         // spawn one more thread than there are elements in the pool
+    //         for _thread_idx in 0..elements_for_pool.len() + 1 {
+    //             threads.push(thread::spawn(move || {
+    //                 pool.acquire();
```

Comment:
The point of this test is to see if the pool behaves correctly when the pool is exhausted.

Right now, your leased item is dropped  right away (before the call to thread::sleep()).
Please fix as follows:

`let _leased_searcher = pool.acquired();`